src/sentry/utils/retries.py

```diff
@@ -47,6 +53,33 @@ def execute_with_retry(*args, **kwargs):
         return decorator
 
 
+class ConditionalRetryPolicy(RetryPolicy[T]):
+    """
+    A basic policy that can be used to retry a callable based on the result
+    of a test function that determines whether or not to retry after the
+    callable throws an exception.
+
+    The test function takes two arguments: the number of times the callable
+    has unsuccesfully been invoked, and the exception instance that was
+    raised during the last execution attempt.
+    """
+
+    def __init__(self, test_function: Callable[[int, Exception], bool]) -> None:
+        self.__test_function = test_function
+
+    def __call__(self, function: Callable[[], T]) -> T:
+        for i in itertools.count(1):
+            try:
+                return function()
+            except Exception as e:
+                if self.__test_function(i, e):
+                    logger.warning(
```

Comment:
I would suggest introduce exponential backoff starting with a few milliseconds. While it seems like ESTALE is not an error that should repeat itself multiple times over in the way we're handling the files, a backoff would further decrease the chance of races here.

Reply:
Added with 18890f026e47ecbfe5f7362bf97eac4f2b4ef7fa:

```
>>> [(i, exponential_delay(0.05)(i)) for i in range(1, 3+1)]
[(1, 0.05), (2, 0.1), (3, 0.2)]
```